### ISSUE TITLE
Question about deep supervision in nnUNetTrainerV2

### ISSUE THREAD

**Issue by hanwinbi** [MIC-DKFZ/nnUNet#329]
Dear Fabian,

Thanks for your hard working on this project. Well, there is some code I can not understand in `nnUNetTrainerV2`(line 78),
```
################# Here we wrap the loss for deep supervision ############
# we need to know the number of outputs of the network
net_numpool = len(self.net_num_pool_op_kernel_sizes)

# we give each output a weight which decreases exponentially (division by 2) as the resolution decreases
# this gives higher resolution outputs more weight in the loss
weights = np.array([1 / (2 ** i) for i in range(net_numpool)])

# we don't use the lowest 2 outputs. Normalize weights so that they sum to 1
mask = np.array([True] + [True if i < net_numpool - 1 else False for i in range(1, net_numpool)])
weights[~mask] = 0
 weights = weights / weights.sum()
self.ds_loss_weights = weights
# now wrap the loss
self.loss = MultipleOutputLoss2(self.loss, self.ds_loss_weights)
 ################# END ###################
```
if the `net_numpool` is 5,
then the weights = [1.     0.5    0.25   0.125  0.0625]
and mask = [True  True  True True False]
so the weights[~mask] = [1.    0.5   0.25  0.125 0.   ]
but actually you mean weights[~mask] = [1.    0.5   0.25  0.    0.   ] right? 
or I understand it wrong.

Best,
Hanwin

**Comment by FabianIsensee** [comment]
Hi Hanwin,
the network has 6 resolution stages. The original and 5 downsampled resolutions. Due to my implementation, the bottleneck of the U-Net has no deep supervision output attached to it, so there is no need to configure a weight for it. Therefore only the last entry in that list is 0.
Best,
Fabian
